workers/reconciliationworker.py

```diff
@@ -0,0 +1,187 @@
+import datetime
+import logging
+import random
+import time
+
+
+import features
+from app import app
+from app import billing as stripe
+from data import marketplace, model
+from data.billing import get_plan
+from data.database import QuaySkuProperties
+from data.model import entitlements
+from util.locking import GlobalLock, LockNotAcquiredException
+from workers.gunicorn_worker import GunicornWorker
+from workers.namespacegcworker import LOCK_TIMEOUT_PADDING
+from workers.worker import Worker
+
+
+logger = logging.getLogger(__name__)
+
+RECONCILIATION_TIMEOUT = 5 * 60  # 5min
+LOCK_TIMEOUT_PADDING = 60  # 60
+RECONCILIATION_FREQUENCY = 5 * 60  # run reconciliation every 5 min
+
+MILLISECONDS_IN_SECONDS = 1000
+SECONDS_IN_DAYS = 86400
+ONE_MONTH = 30 * SECONDS_IN_DAYS * MILLISECONDS_IN_SECONDS
+
+
+class ReconciliationWorker(Worker):
+    def __init__(self):
+        super(ReconciliationWorker, self).__init__()
+        self.add_operation(
+            self._reconcile_entitlements,
+            app.config.get("RECONCILIATION_FREQUENCY", RECONCILIATION_FREQUENCY),
+        )
+
+    def _check_stripe_matches_sku(self, user, sku):
+        """
+        Check if user's stripe plan matches with RH sku
+        """
+        stripe_customer = stripe.Customer.retrieve(user.stripe_id)
+        logger.debug("stripe customer = %s", stripe_customer)
+        if stripe_customer is None:
+            logger.debug("user %s has no valid subscription on stripe", user.username)
+            return False
+
+        plan = get_plan(stripe_customer.subscription.plan.id)
+        if plan["privateRepos"] == sku.value:
+            return True
+
+        logger.debug("stripe plan and sku %s do not match for %s", sku.name, user.username)
+        return False
+
+    def _perform_reconciliation(self, user_api, marketplace_api):
+        """
+        Gather all entitlements from internal marketplace api and store in quay db
+        Create new entitlements for stripe customers if needed
+        """
+        logger.info("Reconciliation worker is gathering entitlements...")
+
+        users = model.user.get_active_users()
```

Comment:
This will loop through every user on each reconciliation, even if no changes were made. Is there any way to do this only when required? If so, does this need to happen both on Stripe change or RH marketplace changes?

Reply:
I'm not sure if there's a way to do it. I think we'll have to keep polling this to make sure entitlements consistently stay in sync. If there was a way, it would only need to happen on RH marketplace changes.